File: api/app/schemas/shipments.py
from pydantic import BaseModel
from typing import List

class ShipmentOverview(BaseModel):
    total_shipments: int
    avg_delivery_time: float
    total_cost: float
    avg_cost_per_shipment: float
    total_weight: float
    unique_origins: int
    unique_destinations: int

class RoutePerformance(BaseModel):
    origin: str
    destination: str
    total_trips: int
    avg_delivery_time: float
    min_delivery_time: float
    max_delivery_time: float
    total_cost: float
    avg_cost: float
    total_weight: float
    unique_vehicles: int
    cost_per_trip: float

class ExpensiveRoute(BaseModel):
    origin: str
    destination: str
    total_shipments: int
    total_cost: float
    avg_cost_per_shipment: float
    avg_delivery_time: float
    vehicles_used: int

class ShipmentAnalytics(BaseModel):
    overview: ShipmentOverview
    expensive_routes: List[ExpensiveRoute]

class PaginatedRouteResponse(BaseModel):
    data: List[RoutePerformance]
    total: int
    page: int
    page_size: int 

File: client/src/types/shipments.ts
export interface ShipmentOverview {
  total_shipments: number;
  avg_delivery_time: number;
  total_cost: number;
  avg_cost_per_shipment: number;
  total_weight: number;
  unique_origins: number;
  unique_destinations: number;
}

export interface RoutePerformance {
  origin: string;
  destination: string;
  total_trips: number;
  avg_delivery_time: number;
  min_delivery_time: number;
  max_delivery_time: number;
  total_cost: number;
  avg_cost: number;
  total_weight: number;
  unique_vehicles: number;
  cost_per_trip: number;
}

export interface ExpensiveRoute {
  origin: string;
  destination: string;
  total_shipments: number;
  total_cost: number;
  avg_cost_per_shipment: number;
  avg_delivery_time: number;
  vehicles_used: number;
}

export interface ShipmentAnalytics {
  overview: ShipmentOverview;
  expensive_routes: ExpensiveRoute[];
}

export interface ShipmentState {
  analytics: {
    data: ShipmentAnalytics | null;
    loading: boolean;
    error: string | null;
  };
  routes: {
    data: RoutePerformance[];
    total: number;
    page: number;
    pageSize: number;
    loading: boolean;
    error: string | null;
  };
}

export interface RouteParams {
  page: number;
  pageSize: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  search?: string;
}

export interface PaginatedRouteResponse {
  data: RoutePerformance[];
  total: number;
  page: number;
  page_size: number;
}